### PR TITLE
fix race condition

### LINKED ISSUE #5534: WARNING: DATA RACE
```console
go run -race . -l ../functional-test/targets-150.txt

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v3.3.1

		projectdiscovery.io

[INF] nuclei-templates are not installed, installing...
[INF] Successfully installed nuclei-templates at /home/runner/nuclei-templates
[INF] Current nuclei version: v3.3.1 (latest)
[INF] Current nuclei-templates version: v9.9.3 (latest)
[WRN] Scan results upload to cloud is disabled.
[INF] New templates added in latest release: 56
[INF] Templates loaded for current scan: 8388
[INF] Executing 8388 signed templates from projectdiscovery/nuclei-templates
[INF] Targets loaded for current scan: 150
[INF] Templates clustered: 1573 (Reduced 221700 Requests)
[INF] Using Interactsh Server: oast.online
==================
WARNING: DATA RACE
Write at 0x00c01590ea70 by goroutine 6[4](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:5)7738:
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).executeRequest()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:783 +0x27d3
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults.func1()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:[5](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:6)17 +0xa74
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:589 +0x78[6](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:7)
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/generic.(*Generic).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/generic/exec.go:61 +0x64c
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec.(*TemplateExecuter).Execute()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/exec.go:199 +0x[7](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:8)c3
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.1()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:139 +0x439
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:146 +0x61

Previous read at 0x00c01590ea70 by goroutine 647[8](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:9)83:
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.wrappedGet()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:230 +0x624
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.Get()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:154 +0x124
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).executeRequest()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:787 +0x285a
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults.func1()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:517 +0xa74
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:58[9](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:10) +0x786
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/generic.(*Generic).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/generic/exec.go:61 +0x64c
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec.(*TemplateExecuter).Execute()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/exec.go:199 +0x7c3
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.1()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:139 +0x439
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:146 +0x61

Goroutine 647738 (running) created at:
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:115 +0x90f
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate.func2()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:129 +0x11c
  github.com/projectdiscovery/hmap/store/disk.(*LevelDB).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/disk/leveldb.go:224 +0x454
  github.com/projectdiscovery/hmap/store/hybrid.(*HybridMap).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/hybrid/hybrid.go:258 +0x1e5
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:137 +0x151
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:80 +0x7f6
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func1()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:135 +0x1[10](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:11)
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:136 +0x41

Goroutine 647883 (running) created at:
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:[11](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:12)5 +0x90f
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate.func2()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:[12](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:13)9 +0x11c
  github.com/projectdiscovery/hmap/store/disk.(*LevelDB).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/disk/leveldb.go:224 +0x454
  github.com/projectdiscovery/hmap/store/hybrid.(*HybridMap).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/hybrid/hybrid.go:258 +0x1e5
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:[13](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:14)7 +0x151
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:80 +0x7f6
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func1()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:135 +0x110
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:136 +0x41
==================
==================
WARNING: DATA RACE
Write at 0x00c00665cb90 by goroutine 709028:
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).executeRequest()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:783 +0x27d3
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults.func1()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:517 +0xa74
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:589 +0x786
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/generic.(*Generic).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/generic/exec.go:61 +0x64c
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec.(*TemplateExecuter).Execute()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/exec.go:199 +0x7c3
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.1()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:139 +0x439
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:[14](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:15)6 +0x61

Previous write at 0x00c00665cb90 by goroutine 709032:
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).executeRequest()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:783 +0x27d3
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults.func1()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:517 +0xa74
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:589 +0x786
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/generic.(*Generic).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/generic/exec.go:61 +0x64c
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec.(*TemplateExecuter).Execute()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/exec.go:199 +0x7c3
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.1()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:139 +0x439
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:146 +0x61

Goroutine 709028 (running) created at:
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:1[15](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:16) +0x90f
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate.func2()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:129 +0x11c
  github.com/projectdiscovery/hmap/store/disk.(*LevelDB).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/disk/leveldb.go:224 +0x454
  github.com/projectdiscovery/hmap/store/hybrid.(*HybridMap).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/hybrid/hybrid.go:258 +0x1e5
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:137 +0x151
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:80 +0x7f6
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func1()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:135 +0x110
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:136 +0x41

Goroutine 709032 (running) created at:
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:115 +0x90f
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate.func2()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:129 +0x11c
  github.com/projectdiscovery/hmap/store/disk.(*LevelDB).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/disk/leveldb.go:224 +0x454
  github.com/projectdiscovery/hmap/store/hybrid.(*HybridMap).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/hybrid/hybrid.go:258 +0x1e5
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:137 +0x151
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:80 +0x7f6
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func1()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:135 +0x110
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:136 +0x41
==================
==================
WARNING: DATA RACE
Write at 0x00c00665cb90 by goroutine 709027:
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).executeRequest()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:783 +0x27d3
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults.func1()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:517 +0xa74
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:589 +0x786
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/generic.(*Generic).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/generic/exec.go:61 +0x64c
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec.(*TemplateExecuter).Execute()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/exec.go:199 +0x7c3
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.1()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:139 +0x439
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:146 +0x61

Previous read at 0x00c00665cb90 by goroutine 709032:
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.(*Configuration).Hash()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:122 +0x284
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.wrappedGet()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:[16](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:17)1 +0x46
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.Get()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:154 +0x124
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).executeRequest()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:787 +0x285a
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults.func1()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:5[17](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:18) +0xa74
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/protocols/http/request.go:589 +0x786
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec/generic.(*Generic).ExecuteWithResults()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/generic/exec.go:61 +0x64c
  github.com/projectdiscovery/nuclei/v3/pkg/tmplexec.(*TemplateExecuter).Execute()
      /home/runner/work/nuclei/nuclei/pkg/tmplexec/exec.go:[19](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:20)9 +0x7c3
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.1()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:139 +0x439
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2.2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:146 +0x61

Goroutine 709027 (running) created at:
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:115 +0x90f
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate.func2()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:129 +0x11c
  github.com/projectdiscovery/hmap/store/disk.(*LevelDB).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/disk/leveldb.go:[22](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:23)4 +0x454
  github.com/projectdiscovery/hmap/store/hybrid.(*HybridMap).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/hybrid/hybrid.go:258 +0x1e5
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:137 +0x151
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:80 +0x7f6
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func1()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:135 +0x110
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:136 +0x41

Goroutine 709032 (running) created at:
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:115 +0x90f
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate.func2()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:129 +0x11c
  github.com/projectdiscovery/hmap/store/disk.(*LevelDB).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/disk/leveldb.go:2[24](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:25) +0x454
  github.com/projectdiscovery/hmap/store/hybrid.(*HybridMap).Scan()
      /home/runner/go/pkg/mod/github.com/projectdiscovery/hmap@v0.0.54/store/hybrid/hybrid.go:[25](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:26)8 +0x1e5
  github.com/projectdiscovery/nuclei/v3/pkg/input/provider/list.(*ListInputProvider).Iterate()
      /home/runner/work/nuclei/nuclei/pkg/input/provider/list/hmap.go:137 +0x151
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateWithTargets()
      /home/runner/work/nuclei/nuclei/pkg/core/executors.go:80 +0x7f6
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func1()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:135 +0x110
  github.com/projectdiscovery/nuclei/v3/pkg/core.(*Engine).executeTemplateSpray.func2()
      /home/runner/work/nuclei/nuclei/pkg/core/execute_options.go:136 +0x41
==================
[INF] Skipped scanme.sh:443 from target list as found unresponsive [30](https://github.com/projectdiscovery/nuclei/actions/runs/10436171980/job/28900900882#step:5:31) times
[INF] Skipped scanme.sh:81 from target list as found unresponsive 36 times
Found 3 data race(s)
exit status 66
```

### PR DESCRIPTION
**Pull request by dogancanbakir**
## Proposed changes

Closes #5534

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [x] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by dogancanbakir** [comment]
perf test passes https://github.com/projectdiscovery/nuclei/actions/runs/10460146436

File: pkg/protocols/http/httpclientpool/clientpool.go [hunk 1]
```diff
@@ -131,6 +148,9 @@ func (c *Configuration) HasStandardOptions() bool {
 
 // GetRawHTTP returns the rawhttp request client
 func GetRawHTTP(options *protocols.ExecutorOptions) *rawhttp.Client {
+	rawHttpClientMu.Lock()
```
Comment (Mzack9999) [inline pkg/protocols/http/httpclientpool/clientpool.go:151]: What do you think about using https://pkg.go.dev/sync#OnceValue to obtain the default client once in a safe fashion?

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
I've tried to execute locally and it seems like we have other few race conditions, for example:

```console
==================
WARNING: DATA RACE
Write at 0x000111024d50 by goroutine 51:
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/protocolstate.Close()
      /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/common/protocolstate/state.go:217 +0x84
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/protocolinit.Close()
      /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/common/protocolinit/init.go:41 +0x1c4
  github.com/projectdiscovery/nuclei/v3/internal/runner.(*Runner).Close()
      /Users/user/go/src/github.com/projectdiscovery/nuclei/internal/runner/runner.go:380 +0x1bc
  main.main.func4()
      /Users/user/go/src/github.com/projectdiscovery/nuclei/cmd/nuclei/main.go:165 +0x11c

Previous read at 0x000111024d50 by goroutine 71255:
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.wrappedGet()
      /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:236 +0x764
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.Get()
      /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:154 +0x124
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).executeRequest()
      /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/http/request.go:787 +0x28da
  github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults.func1()
```

I think we should also consider in a follow up issue to disable the panic-recover mechanism in GitHub Actions, where not necessary. Otherwise, errors are not detected. The app should just crash to ensure failures are visible.

File: pkg/protocols/http/httpclientpool/clientpool.go (hunk 1 above)
Comment (dogancanbakir) [reply pkg/protocols/http/httpclientpool/clientpool.go]: sync.OnceValue expects niladic func, thus making things a bit more complex. I used sync.Once instead.

**Comment by dogancanbakir** [comment]
> I've tried to execute locally and it seems like we have other few race conditions, for example:
> 
> ```
> ==================
> WARNING: DATA RACE
> Write at 0x000111024d50 by goroutine 51:
>   github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/protocolstate.Close()
>       /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/common/protocolstate/state.go:217 +0x84
>   github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/protocolinit.Close()
>       /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/common/protocolinit/init.go:41 +0x1c4
>   github.com/projectdiscovery/nuclei/v3/internal/runner.(*Runner).Close()
>       /Users/user/go/src/github.com/projectdiscovery/nuclei/internal/runner/runner.go:380 +0x1bc
>   main.main.func4()
>       /Users/user/go/src/github.com/projectdiscovery/nuclei/cmd/nuclei/main.go:165 +0x11c
> 
> Previous read at 0x000111024d50 by goroutine 71255:
>   github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.wrappedGet()
>       /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:236 +0x764
>   github.com/projectdiscovery/nuclei/v3/pkg/protocols/http/httpclientpool.Get()
>       /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/http/httpclientpool/clientpool.go:154 +0x124
>   github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).executeRequest()
>       /Users/user/go/src/github.com/projectdiscovery/nuclei/pkg/protocols/http/request.go:787 +0x28da
>   github.com/projectdiscovery/nuclei/v3/pkg/protocols/http.(*Request).ExecuteWithResults.func1()
> ```
> 
> I think we should also consider in a follow up issue to disable the panic-recover mechanism in GitHub Actions, where not necessary. Otherwise, errors are not detected. The app should just crash to ensure failures are visible.

I couldn't repro this, but I removed the redundant nil assignment here: https://github.com/projectdiscovery/nuclei/pull/5547/commits/35a0d673ad8e12b11e90e8e0090feb26ea042b46. Could you re-run and see if it happens again?

File: pkg/protocols/common/protocolstate/state.go [hunk 2]
```diff
@@ -214,6 +214,5 @@ func Close() {
 		Dialer.Close()
 		Dialer = nil
 	}
-	Dialer = nil
```
Comment (Mzack9999) [inline pkg/protocols/common/protocolstate/state.go:217]: I think this might be wanted to release the global instance of the Dialer

Comment (Mzack9999) [reply pkg/protocols/common/protocolstate/state.go]: Disregard my comment, missed the previous assignment

**Pull request merged**